Enhance remote logging data privacy

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Related to https://github.com/woocommerce/woocommerce/issues/52037

This PR updates the remote logging feature to address privacy concerns. The changes are:

1. Only log errors caused by WooCommerce, rather than logging all errors containing a WooCommerce file in the trace.
2. Redact potential user data such as email addresses, phone numbers, credit card number and IP.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


1. Create a fresh site using this branch
2. Go to `/wp-admin/tools.php?page=woocommerce-admin-test-helper` > `Remote Logging` and enable remote logging
3. Modify [`load_rest_api`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-woocommerce.php#L538-L541) to throw an exception with sensitive data

```php
throw new Exception( 'Test exception with real sensitive data: john@example.com 555-123-4567 192.168.1.1 1234 5678 9012 3456' );
```

4. Go to Logstash page PCYsg-11Xv-p2 and confirm that the sensitive data is redacted

`Uncaught Exception: Test exception with real sensitive data: [redacted_email] [redacted_phone] [redacted_ip] [redacted_credit_card] in ./woocommerce/includes/class-woocommerce.php....`

5. Revert the changes to `load_rest_api`
6. On remote logging page, Reset the PHP rate limit
7. Modify [`woocommerce-beta-tester.php`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-beta-tester/woocommerce-beta-tester.php#L132-L141) to throw an exception that is not directly caused by WooCommerce but contains WooCommerce files in the trace

```php
function test_throw_error() {
	throw new Exception( 'Test PHP exception from WooCommerce Beta Tester' );
}

add_action(
	'before_woocommerce_init',
	function() {
		if ( ! get_option( 'wc_simulated_error' ) ) {
			add_option( 'wc_simulated_error', 'yes' );
			test_throw_error();
		}

		if ( class_exists( '\Automattic\WooCommerce\Utilities\FeaturesUtil' ) ) {
			\Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, true );
			\Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'product_block_editor', __FILE__, true );
		}
	}
);
```

8. Confirm that the error is not logged

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Refactor RemoteLogger to improve backtrace handling and sanitize sensitive data

chihsuan committed:
Add changelog

chihsuan committed:
Fix remote logger

chihsuan committed:
Refactor RemoteLoggerTest to improve error handling and sanitize sensitive data

rjchow commented on plugins/woocommerce/src/Internal/Logging/RemoteLogger.php:
```diff
@@ -470,6 +485,26 @@ function ( $trace_item ) {
+	 * @param string $message The message to redact.
+	 * @return string The redacted message.
+	 */
+	private function redact_user_data( $message ) {
```
I'm wondering if we should make this a filter so that we can add more arbitrary sanitisation? We'd have to make sure that the wp filter code has been loaded by this point though

rjchow replied:
If we're sanitising PII we should probably include credit card numbers too, since those are quite likely to pop up in woocommerce context

chihsuan replied:
That's a good idea! I'll add that.

chihsuan replied:
Added filter and updated the method to sanitise credit card numbers as well. 0100bf4a4f8bd890a2929386e7acdbe8d2a03c6e

rjchow commented on plugins/woocommerce/src/Internal/Logging/RemoteLogger.php:
```diff
@@ -470,6 +485,26 @@ function ( $trace_item ) {
+		$message = preg_replace( '/[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/', '[redacted_email]', $message );
+
+		// Redact potential phone numbers (this is a simple pattern and might need refinement).
+		$message = preg_replace( '/\b\d{3}[-.]?\d{3}[-.]?\d{4}\b/', '[redacted_phone]', $message );
```
let's make this more generic to include non-US number formats, what do you think of something like
`/\+?(\(?\d{1,3}\)?)?(([-.\s]?\d{2,4}){1,4}|\d{5,10})/` 

Explanation

	1.	\+?: Matches an optional + at the start.
	2.	(\(?\d{1,3}\)?)?: Matches an optional area code, which:
	•	\(?: Allows an optional opening parenthesis (.
	•	\d{1,3}: Matches 1 to 3 digits for the area code.
	•	\)?: Allows an optional closing parenthesis ).
	3.	(([-.\s]?\d{2,4}){1,4}|\d{5,10}): Matches either:
	•	([-.\s]?\d{2,4}){1,4}: 1 to 4 groups, each containing 2 to 4 digits, optionally separated by -, ., or whitespace.
	•	|: OR
	•	\d{5,10}: A consecutive string of 5 to 10 digits.

Example Matches

This regex will match patterns like:

	•	+1 (123) 456 7890 (with parentheses around area code)
	•	+44-123-4567-890 (with area code, no parentheses)
	•	1234567890 (10 consecutive digits, no area code)
	•	(123) 456-7890 (area code in parentheses, groups)
	•	+91 12345 (5 consecutive digits, no parentheses)


Phone number match is [notoriously difficult](https://github.com/google/libphonenumber), but this is a good start and thanks for making the effort!

chihsuan replied:
Thanks for the suggestion!  I'll let you know once I've made the changes.

chihsuan replied:
I've updated the regex to include more number formats and added more test cases to ensure it covers a wide range of phone number variations. 0100bf4a4f8bd890a2929386e7acdbe8d2a03c6e

I didn't use the provided regex because it also matches the number like `123`. I asked ChatGPT for confirmation on this.

chihsuan committed:
Address PR feedback

- Add a filter so that we can add more arbitrary sanitisation
- Update phone number sanitisation
- Add credit card sanitisation

Fix remote logger

chihsuan committed:
Fix lint

psealock commented:
@chihsuan This is looking great.

>Modify [load_rest_api](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-woocommerce.php#L538-L541) to throw an exception with sensitive data

When I modify these lines, I get an error in the UI but nothing shows up in LogStash.

<img width="824" alt="image" src="https://github.com/user-attachments/assets/0db35405-5752-4532-9d5f-85f9d56e2431">

Just to be sure, I fired an error from the Beta Tester and that Core exception was properly logged in LogStash. Can you reproduce?

chihsuan commented:
> Just to be sure, I fired an error from the Beta Tester and that Core exception was properly logged in LogStash. Can you reproduce?

I was able to reproduce this on JN site. Looks like the file path is sanitized to an invalid value. I'll fix it.

chihsuan committed:
Merge branch 'trunk' into enhance/remote-logging-data-privacy

chihsuan committed:
Fix sanitize file path in error context

chihsuan commented:
@psealock Fixed in https://github.com/woocommerce/woocommerce/pull/52321/commits/cfb8445d3aea9f8b953db4ee648add83d9788543.

chihsuan committed:
Merge branch 'trunk' into enhance/remote-logging-data-privacy

chihsuan commented:
> @chihsuan , this is looking good! Could we redact API keys as part of this too or would it be out of scope? I quickly generated a few patterns with chatgpt
> 
> ```
> $patterns = [
>         '/\b[A-Za-z0-9]{32,40}\b/',                // Generic API key
>         '/\b[0-9a-f]{32}\b/i',                     // 32 hex characters
>         '/\b(?:[A-Z0-9]{4}-){3,7}[A-Z0-9]{4}\b/i', // Segmented API key (e.g., XXXX-XXXX-XXXX-XXXX)
>         '/\bsk_[A-Za-z0-9]{24,}\b/i',              // Stripe keys (starts with sk_)
>     ];
> ```
Sure, thanks for the suggestion! I'll add that.

chihsuan committed:
Redact API keys

chihsuan commented:
Ok, I've added the patterns to redact API keys in 231825703a8e116c6e6ec9a81c9e996d7bbacf2a

chihsuan committed:
Fix tests

psealock reviewed (approved):
This is working well, sensitive info was correctly redacted and non WooCommerce errors weren't logged.

The phone number thing is indeed tricky. This won't cover all scenarios, but good to get the bulk of them.

rjchow reviewed (approved):
Thank you! +1000 approve

adrianduffell commented:
> Ok, I've added the patterns to redact API keys in [2318257](https://github.com/woocommerce/woocommerce/commit/231825703a8e116c6e6ec9a81c9e996d7bbacf2a)

Nice, thanks @chihsuan !
